### ISSUE TITLE
"InvalidAsn1Error: encoding too long" after updating to 2.8.2

### ISSUE THREAD

**Issue by oniemela** [tngan/samlify#452]
I recently wrote about another issue (#449 ), and after updating from 2.7.7 to the latest 2.8.2, we are still experiencing issues

Same works fine with previous version, but we need to update for the security patches that are in later releases

Call:
serviceProvider.createLoginRequest(identityProvider, 'redirect', (logoutRequestTemplate) => { ... })

Failure exception:
Exception: InvalidAsn1Error: encoding too long
Stack: InvalidAsn1Error: encoding too long
    at newInvalidAsn1Error (C:\...\node_modules\asn1\lib\ber\errors.js:7:13)
    at Reader.readLength (C:\...\node_modules\asn1\lib\ber\reader.js:103:13)
    at Reader.readSequence (C:\...\node_modules\asn1\lib\ber\reader.js:136:16)
    at Object.privateImport (C:\...\node_modules\node-rsa\src\formats\pkcs1.js:68:16)
    at Object.detectAndImport (C:\...\node_modules\node-rsa\src\formats\formats.js:64:48)     
    at NodeRSA.module.exports.NodeRSA.importKey (C:\...\node_modules\node-rsa\src\NodeRSA.js:183:22)
    at new NodeRSA (C:\...\node_modules\node-rsa\src\NodeRSA.js:73:18)
    at Object.constructMessageSignature (C:\...\node_modules\samlify\build\src\libsaml.js:453:32)
    at buildRedirectURL (C:\...\node_modules\samlify\build\src\binding-redirect.js:82:80)     
    at Object.loginRequestRedirectURL (C:\...\node_modules\samlify\build\src\binding-redirect.js:121:22).

The issue stems from this call:
https://github.com/tngan/samlify/blob/b422c21a6162f3ad6d3ffa527a775d3f0edd4e3d/src/libsaml.ts#L547
```
const decryptedKey = new nrsa(
        utility.readPrivateKey(key, passphrase),
        'private',
        {
          signingScheme: getSigningScheme(signingAlgorithm),
        }
      );
```

In Node-RSA documentation, its stated it should be called like (note the array around key and format):
`const key = new NodeRSA([keyData, [format]], [options]);`
(https://github.com/rzcoder/node-rsa)

but as it is, it seems to be called:
`const key = new NodeRSA(keyData, [format], [options]);`

Not sure if this is a problem with the Node-RSA documentation not being correct, we tested this by changing the call as
```
var decryptedKey = new node_rsa_1.default([utility_1.default.readPrivateKey(key, passphrase), 'private'], {
                signingScheme: getSigningScheme(signingAlgorithm),
            });
```

The call seemed to work with this change, but I'm not sure about the repercussions (if its actually creating a private-format decrypted key or not (or if the problem is from somewhere else)
We use @authenio/samlify-xsd-schema-validator and Azure functions typescript implementation (typescript version 4.0.3)

Thank you very much for your time in advance

**Comment by tngan** [comment]
@oniemela I have checked the test file, the second argument is the key format in string. See if you can try different formats, I wonder forcing to use `private` might not be a good option, probably need to expose as an option instead.

```ts
// https://github.com/rzcoder/node-rsa/blob/e7e7f7d2942a3bac1d2e132a881e5a3aceda10a1/test/tests.js#L375
let key = new NodeRSA(privateFileNoBoundaries, "pkcs1-private-pem");
```

**Comment by parole123** [comment]
@tngan please check again, should be bug? samlify is passing 3 arguments?

**Comment by jrasanen** [comment]
I have this issue when generating a logout request.

@oniemela, node-rsa's documentation has a typo, it should be three parameters.

By passing an array and an object `new NodeRSA([utility.readPrivateKey(key, passphrase), 'private'], {...})` it generates a new keypair as a side-effect for it. 

It goes to [this highlighted branch of code in node-rsa](https://github.com/rzcoder/node-rsa/blob/e7e7f7d2942a3bac1d2e132a881e5a3aceda10a1/src/NodeRSA.js#L74-L76) and checks if the `key` parameter is an object which in our case is the array and executes it as following

```javascript
const isObject = function (value) {
   var type = typeof value;
   return !!value && (type == 'object' || type == 'function');
}

isObject([utility.readPrivateKey(key, passphrase), 'private']) // true
```

Then generateKeyPair gets two null parameters as a result,
```
this.generateKeyPair(key.b, key.e);
```
generateKeyPair has [two optional parameters](https://github.com/rzcoder/node-rsa/blob/e7e7f7d2942a3bac1d2e132a881e5a3aceda10a1/src/NodeRSA.js#L157-L158), and it defaults to 2048, 65537 if both are null

---

So to summarise, Samlify definitely should pass 3 arguments, otherwise we are generating a new keypair as an unwanted side-effect, but there is some other issue related to this.

**Comment by jrasanen** [comment]
I solved this by using a private key that is password protected,

> openssl genrsa -passout pass:foobar -out encryptKey.pem 4096
> openssl req -new -x509 -key encryptKey.pem -out encryptionCert.cer -days 3650

Private key without a password throws the `InvalidAsn1Error`

**Comment by ekrister** [comment]
Ran into this bug also when trying to update modules. Maybe this is obvious, but I just wanted to add, that the bug happens when updating node-rsa module from 1.0.5 to 1.1.1.

**Comment by parole123** [comment]
PKCS8 to PKCS1
-----BEGIN PRIVATE KEY----- >>> -----BEGIN RSA PRIVATE KEY-----

`openssl pkcs8 -in pk8key.pem -traditional -nocrypt -out key.pem`

**Comment by ekrister** [comment]
> 

Thank you, this solved my issue!

**Comment by khierhol** [comment]
Same issue as for OP with "". Some additional info:
a. it is dependent on the situation / actual credentials/payload, on a identical configured test machine with it does not occur
b. the error is dependent on the samlify version, it occurs exactly when going from 2.7.7 (ok) to 2.8.0 and then occurs for any version from 2.8.0 on
